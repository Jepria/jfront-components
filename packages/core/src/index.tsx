export * from '@jfront-ui/app'
export * from '@jfront-ui/datepicker'
export * from '@jfront-ui/form'
//export * from '@jfront-ui/grid'
export * from '@jfront-ui/hooks'
//export * from '@jfront-ui/input'
export * from '@jfront-ui/layout'
//export * from '@jfront-ui/pagingbar'
//export * from '@jfront-ui/tabpanel'
//export * from '@jfront-ui/toolbar'
