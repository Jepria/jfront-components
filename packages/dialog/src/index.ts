export * from './SimpleDialog'
export * from './AlertDialog'
export * from './FormDialog'
export * from './ErrorDialog'
export * from './WarningDialog'