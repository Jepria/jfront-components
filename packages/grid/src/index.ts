export * from './common'
export * from './jepgrid'
