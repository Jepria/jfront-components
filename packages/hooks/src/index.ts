export * from "./useOnClickOutside"
export * from "./useRouteMapping"
export * from "./useTree/"
export * from "./useTree/types"
export * from "./useFilter"
