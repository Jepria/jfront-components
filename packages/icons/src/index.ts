export * from "./Add"
export * from "./Delete"
export * from "./Edit"
export * from "./Exclamation"
export * from "./First"
export * from "./Last"
export * from "./Loading"
export * from "./Next"
export * from "./OpenIcon"
export * from "./Prev"
export * from "./Refresh"
export * from "./Save"
export * from "./Search"
export * from "./SettingsSmall"
export * from "./Split"
export * from "./View"
