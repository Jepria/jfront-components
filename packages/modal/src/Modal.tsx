import React, { createContext, useContext, useEffect } from "react"
import {
  GlassMask,
  Container,
  StyledDialog,
  Header,
  CloseIcon,
  Button,
  Content,
  Footer,
} from "./styles"
import { forwardRef, isFunction } from "@jfront/ui-utils"

export type UseModal = {
  visible?: boolean
  onOpen?: () => void
  onClose?: () => void
}

const UseModalContext = createContext<UseModal>({})

export const useModal = () => {
  const context = useContext(UseModalContext)
  return context
}

const cloneChildren = (children: React.ReactNode, props: any) => {
  if (isFunction(children)) {
    return children(props)
  }
  return React.Children.map(children, (child) => {
    if (React.isValidElement(child)) {
      return React.cloneElement(child, { ...props })
    } else {
      return child
    }
  })
}

export interface ModalProps extends React.HTMLAttributes<HTMLDivElement> {
  visible?: boolean
  closeOnOutsideClick?: boolean
  onOpen?: () => void
  onClose?: () => void
  children: React.ReactNode
}

export const Modal = forwardRef<ModalProps, "div">(
  ({ visible, onClose, onOpen, closeOnOutsideClick, as, ...props }, ref) => {
    useEffect(() => {
      if (visible && onOpen) {
        onOpen()
      }
    }, [onOpen, visible])

    if (as) {
      const Component = as
      return (
        <>
          {visible && (
            <>
              <GlassMask />
              <Container onClick={closeOnOutsideClick ? onClose : undefined}>
                <UseModalContext.Provider
                  value={{
                    visible,
                    onClose,
                    onOpen,
                  }}
                >
                  <Component {...props} onClose={onClose} ref={ref}>
                    {cloneChildren(props.children, { onClose })}
                  </Component>
                </UseModalContext.Provider>
              </Container>
            </>
          )}
        </>
      )
    } else {
      return (
        <>
          {visible && (
            <>
              <GlassMask />
              <Container onClick={closeOnOutsideClick ? onClose : undefined}>
                <UseModalContext.Provider
                  value={{
                    visible,
                    onClose,
                    onOpen,
                  }}
                >
                  <StyledDialog {...props} ref={ref}>
                    {cloneChildren(props.children, { onClose })}
                  </StyledDialog>
                </UseModalContext.Provider>
              </Container>
            </>
          )}
        </>
      )
    }
  },
)

export interface ModalCloseButton {
  tooltip?: string
}

export const ModalCloseButton = (props: ModalCloseButton) => {
  const { onClose } = useModal()
  return (
    <Button title={props.tooltip}>
      <CloseIcon onClick={onClose} />
    </Button>
  )
}

export interface ModalHeaderProps extends React.HTMLAttributes<HTMLDivElement> {
  withCloseButton?: boolean
  children?: React.ReactNode
}

export const ModalHeader = forwardRef<ModalHeaderProps, "header">(
  ({ as, withCloseButton, ...props }, ref) => {
    const { onClose } = useModal()

    if (as) {
      const Component = as
      return (
        <>
          <Component {...props} onClose={onClose} ref={ref}>
            {cloneChildren(props.children, { onClose })}
          </Component>
          {withCloseButton && <ModalCloseButton />}
        </>
      )
    } else {
      return (
        <Header {...props} style={withCloseButton ? {...props.style, paddingRight: "16px"} : {...props.style}} ref={ref}>
          {cloneChildren(props.children, { onClose })}
          {withCloseButton && <ModalCloseButton />}
        </Header>
      )
    }
  },
)

export interface ModalContentProps
  extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode
}

export const ModalContent = forwardRef<ModalContentProps, "section">(
  ({ as, ...props }, ref) => {
    const { onClose } = useModal()

    if (as) {
      const Component = as
      return (
        <Component {...props} onClose={onClose} ref={ref}>
          {cloneChildren(props.children, { onClose })}
        </Component>
      )
    } else {
      return (
        <Content {...props} ref={ref}>
          {cloneChildren(props.children, { onClose })}
        </Content>
      )
    }
  },
)

export interface ModalFooterProps extends React.HTMLAttributes<HTMLDivElement> {
  onClose?: () => void
  children: React.ReactNode
}

export const ModalFooter = forwardRef<ModalFooterProps, "footer">(
  ({ as, ...props }, ref) => {
    const { onClose } = useModal()

    if (as) {
      const Component = as
      return (
        <Component {...props} onClose={onClose} ref={ref}>
          {cloneChildren(props.children, { onClose })}
        </Component>
      )
    } else {
      return (
        <Footer {...props} ref={ref}>
          {cloneChildren(props.children, { onClose })}
        </Footer>
      )
    }
  },
)
