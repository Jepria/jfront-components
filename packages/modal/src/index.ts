export * from "./Modal"
export { GlassMask } from './styles';
