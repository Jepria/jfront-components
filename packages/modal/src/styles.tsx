import styled from "styled-components"
import { CloseImage } from "@jfront/ui-icons"

export const Header = styled.header`
  background: ${(props) => props.theme.modal.header.bgColor};
  padding: 5px 10px;
  margin: 0;
  color: ${(props) => props.theme.modal.header.color};
  border-top-left-radius: ${(props) => props.theme.modal.header.borderRadius};
  border-top-right-radius: ${(props) => props.theme.modal.header.borderRadius};
  min-height: 18px;
  font-weight: bold;
  font-size: ${(props) => props.theme.fontSize.md};
  font-family: ${(props) => props.theme.fontFamily};
`
Header.defaultProps = {
  theme: {
    fontFamily: "tahoma, arial, helvetica, sans-serif",
    fontSize: {
      md: "12px",
    },
    modal: {
      header: {
        borderWidth: 0,
        borderStyle: "",
        borderColor: "",
        borderRadius: "5px",
        bgColor: "linear-gradient(rgb(255, 255, 255), rgb(208, 222, 240))",
        color: "rgb(21, 66, 139)",
      },
    },
  },
}

export const GlassMask = styled.div`
  top: 0;
  left: 0;
  position: absolute;
  height: 100vh;
  width: 100vw;
  z-index: 1100;
  opacity: 0.2;
  background-color: black;
`

export interface StyledDialogProps {
  type?: string
  x?: number
  y?: number
}

const defaultProps = {
  theme: {
    modal: {
      borderWidth: 0,
      borderStyle: "solid",
      borderColor: "#ccc",
      borderRadius: "5px",
      bgColor: "#fff",
    },
  },
}

export const StyledDraggableView = styled.div`
  position: relative;
  z-index: 1102;
  box-sizing: border-box;
  overflow-wrap: break-word;
  max-width: 400px;
  background: ${(props) => props.theme.modal.bgColor};
  border-radius: ${(props) => props.theme.modal.borderRadius};
  display: inline-flex;
  flex-direction: column;
  opacity: 0.75;
  box-shadow: 4px 4px 8px 0px rgb(34, 60, 80);
`
StyledDraggableView.defaultProps = defaultProps

export const StyledDialog = styled.div<StyledDialogProps>`
  position: relative;
  z-index: 1102;
  box-sizing: border-box;
  overflow-wrap: break-word;
  max-width: 400px;
  background: ${(props) => props.theme.modal.bgColor};
  border-radius: ${(props) => props.theme.modal.borderRadius};
  display: inline-flex;
  flex-direction: column;
  left: ${(props) => (props.x ? `${props.x}px` : "50%")};
  top: ${(props) => (props.y ? `${props.y}px` : "50%")};
  transform: translate(-50%, -50%);
`
StyledDialog.defaultProps = defaultProps

export const Content = styled.section`
  box-sizing: border-box;
  width: 100%;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  overflow-y: auto;
  padding: 5px;
`

export const Footer = styled.footer`
  box-sizing: border-box;
  width: 100%;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  padding: 5px;
`

export const StyledContainer = styled.div`
  top: 0;
  left: 0;
  position: absolute;
  z-index: 1101;
  width: 100vw;
  height: 100vh;
`

export const Button = styled.span`
  z-index: 1103;
  position: absolute;
  top: 1px;
  right: 1px;
  color: ${(props) => props.theme.modal.closeButtonColor};
`

Button.defaultProps = {
  theme: {
    modal: {
      closeButtonColor: "rgb(21, 66, 139)",
    },
  },
}

export const CloseIcon = styled(CloseImage)`
  height: 16px;
  width: 16px;
  fill: currentColor;
  &:hover {
    opacity: 0.5;
  }
`

export const StyledLayer = styled.div`
  width: 100vw;
  height: 100vh;
  position: absolute;
  top: 0;
  left: 0;
  pointer-events: none;
  z-index: 1105;
`

export const DraggableBlock = styled.div`
  display: inline-block;
  box-sizing: border-box;
  padding: 0;
  margin: 0;
`
