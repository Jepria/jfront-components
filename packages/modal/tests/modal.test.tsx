import { render } from "@testing-library/react"
import React from "react"
import { Modal } from "../src"

test("Checking for the existence of an element Modal", () => {
  expect(render(<Modal visible>Text</Modal>))
})

test("Matches snapshot", () => {
  const { asFragment } = render(<Modal visible>Text</Modal>)
  expect(asFragment()).toMatchSnapshot()
})