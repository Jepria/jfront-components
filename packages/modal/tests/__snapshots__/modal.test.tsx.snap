// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-AxhCb fJYEYh"
  />
  <div
    class="sc-Axmtr eBZWUw"
  >
    <div
      class="sc-AxhUy dybvAb"
    >
      Text
    </div>
  </div>
</DocumentFragment>
`;
